reqh/reqh_service.c

```diff
@@ -953,17 +967,8 @@ M0_INTERNAL void m0_reqh_service_ctxs_shutdown_prepare(struct m0_reqh *reqh)
 		}
 		reqh_service_ctx_sm_unlock(ctx);
 		/* Do waiting outside the sm lock. */
-		if (connecting) {
-			m0_clink_add_lock(&ctx->sc_rlink.rlk_wait,
-					  &ctx->sc_rlink_abort);
-			/*
-			 * While we were leaving the sm lock the fom activity
-			 * might happen alright, thus timed waiting.
-			 */
-			m0_chan_timedwait(&ctx->sc_rlink_abort,
-				m0_time_from_now(REQH_SVC_CONNECT_TIMEOUT, 0));
-			m0_clink_del_lock(&ctx->sc_rlink_abort);
-		}
+		if (connecting)
+			wait_connecting(ctx);
```

Comment:
I wonder if we could just call `reqh_service_ctx_state_wait(ctx, M0_RSC_OFFLINE)` here. As far as I understand, the goal of this code is to wait until rpc link FOMs are done their work. If we wait until reqh_service_ctx_ast_cb() leaves the state machine in OFFLINE state then we can be sure there is no rlink-related activity after this point (assuming the `rlk_rc = M0_ERR(-EPERM)` trick works fine).

Reply:
Done.